### PR TITLE
Add post about memkind-dax-kmem

### PR DESCRIPTION
**Pull request by michalbiesek**
**Please do not merge** before:
- [x] Memkind release 1.10.0
- [ ] Previous post regarding memkind #118 is merged

Reviews of this PR are highly welcome
Rendered page is available on ~~https://michalbiesek.github.io/2019/12/31/memkind-dax-kmem.html~~
https://michalbiesek.github.io/2020/01/20/memkind-dax-kmem.html

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/pmem.github.io/120)

### CODE REVIEW

File: _posts/2019-12-31-memkind-dax-kmem.md [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+layout: post
+identifier: memkind-support-for-KMEM-DAX-option
+---
+
+### Introduction
+
+Linux kernel version 5.1 brings in [support][kernel-patch] for the volatile-use of persistent memory
+as a hotplugged memory region (**KMEM DAX**).
+From the perspective of OS, persistent memory will be seen as a separate NUMA node(s).
+[libmemkind][memkind-release] extends its API by new kinds which allow for automatic detecting and allocating to
```
Comment (pbalcer) [inline _posts/2019-12-31-memkind-dax-kmem.md]: automatic detection and allocation from these new Persistent Memory NUMA nodes.

File: _posts/2019-12-31-memkind-dax-kmem.md [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+			<M>     PMEM DAX: direct access to persistent memory
+			<M>     KMEM DAX: volatile-use of persistent memory
+			< >   PMEM DAX: support the deprecated /sys/class/dax interface
+{% endhighlight %}
+
+[ndctl and daxctl][ndctl-release] with version 66 or later.
+
+### System Configuration
+
+Using NVDIMM as an extension of DRAM needs proper configuration with `ndctl` and `daxctl`.
```
Comment (pbalcer) [inline _posts/2019-12-31-memkind-dax-kmem.md]: ... needs to be configured with ndctl and daxctl.

File: _posts/2019-12-31-memkind-dax-kmem.md [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+A Reconfiguration of Device-DAX depends on dax-bus device model.
```
Comment (pbalcer) [inline _posts/2019-12-31-memkind-dax-kmem.md]: drop 'a'

File: _posts/2019-12-31-memkind-dax-kmem.md [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+Kernel should support `/sys/bus/dax model`. To migrate it from
+`sys/class/dax` to `/sys/bus/dax model` please use [daxctl-migrate-device-mode][daxctl-migrate-device-model].
+
+### Reconfigure dax devices
+
+To see the list of available NUMA nodes on the system you could use `numactl`.
```
Comment (pbalcer) [inline _posts/2019-12-31-memkind-dax-kmem.md]: The list of available NUMA nodes on the system can be retrieved using `numactl`.

File: _posts/2019-12-31-memkind-dax-kmem.md [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+]
+{% endhighlight %}
+
+To [reconfigure][daxctl-reconfigure-device] DAX device from `devdax` mode to a `system-ram-mode`:
+
+{% highlight console %}
+$ daxctl reconfigure-device dax1.0 --mode=system-ram
+{% endhighlight %}
+
+With this operation persistent memory could be seen as separate NUMA and be used as volatile RAM.
```
Comment (pbalcer) [inline _posts/2019-12-31-memkind-dax-kmem.md]: After this operation, persistent memory will be configured as separate NUMA node and can be used as volatile memory.

File: _posts/2019-12-31-memkind-dax-kmem.md [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+For configuration below persistent memory NUMA node is Node 3 (NUMA node to which does not belong any logical CPU)
```
Comment (pbalcer) [inline _posts/2019-12-31-memkind-dax-kmem.md]: For the example configuration below, persistent memory ...

File: _posts/2019-12-31-memkind-dax-kmem.md [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+Note:
+- Image is simplified to present only 2 CPUs: CPU 0 which belong to Node 0 and CPU 30 which belong to Node 1
+- Numbers in squares corresponds to NUMA node distances between Nodes 0/1 and all of the Nodes
+- Node 0 and Node 1 are based on DRAM memory where Node 2 and Node 3 are based on NVDIMM memory
+- Moving physical memory to swap space is limited only to one Node
+- To simplify examples below we assume that process uses only CPU 0
+
+**MEMKIND_DAX_KMEM**:
+This is the first variant where the allocation comes from the closest persistent memory NUMA node
+at the time of allocating.
```
Comment (pbalcer) [inline _posts/2019-12-31-memkind-dax-kmem.md]: at the time of allocation.

File: _posts/2019-12-31-memkind-dax-kmem.md [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+![memkind_dax_kmem](/assets/memkind_dax_kmem.gif)
+
+The process runs only CPU 0, CPU which is assigned to Node 0. Node 2 is the closest persistent memory NUMA node to Node 0,
+therefore using **MEMKIND_DAX_KMEM** results in taking memory only from Node 2. If there is not enough free memory in Node 2
+to satisfy an allocation request, inactive pages from Node 2 are moved into the swap space - freeing up the memory for
+**MEMKIND_DAX_KMEM** use case.
+
+**MEMKIND_DAX_KMEM_ALL**:
+This is the second variant where the allocation comes from the closest persistent memory NUMA node
+available at the time allocating.
```
Comment (pbalcer) [inline _posts/2019-12-31-memkind-dax-kmem.md]: allocation

File: _posts/2019-12-31-memkind-dax-kmem.md [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+
+![memkind_dax_kmem_all](/assets/memkind_dax_kmem_all.gif)
+
+A similar situation to scenario presented in **MEMKIND_DAX_KMEM** except that when there is not enough free
+memory in Node 2 to satisfy the allocation request, an allocation pattern switch to Node 3. When available free space is
+exhausted in Node 3 - swap space is used.
+
+**MEMKIND_DAX_KMEM_PREFERRED**:
+This is the third variant where the allocation comes from the closest persistent memory NUMA node at the time
+of allocating. If there is not enough memory to satisfy the request the allocation will fall back on other memory
```
Comment (pbalcer) [inline _posts/2019-12-31-memkind-dax-kmem.md]: allocation

File: _posts/2019-12-31-memkind-dax-kmem.md [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+Note:
+- Image is simplified to present only 2 CPUs: CPU 0 which belong to Node 0 and CPU 30 which belong to Node 1
+- Numbers in squares corresponds to NUMA node distances between Nodes 0/1 and all of the Nodes
+- Node 0 and Node 1 are based on DRAM memory where Node 2 and Node 3 are based on NVDIMM memory
+- Moving physical memory to swap space is limited only to one Node
+- To simplify examples below we assume that process uses only CPU 0
+
+**MEMKIND_DAX_KMEM**:
+This is the first variant where the allocation comes from the closest persistent memory NUMA node
```
Comment (pbalcer) [inline _posts/2019-12-31-memkind-dax-kmem.md]: where the memory

File: _posts/2019-12-31-memkind-dax-kmem.md [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+
+![memkind_dax_kmem](/assets/memkind_dax_kmem.gif)
+
+The process runs only CPU 0, CPU which is assigned to Node 0. Node 2 is the closest persistent memory NUMA node to Node 0,
+therefore using **MEMKIND_DAX_KMEM** results in taking memory only from Node 2. If there is not enough free memory in Node 2
+to satisfy an allocation request, inactive pages from Node 2 are moved into the swap space - freeing up the memory for
+**MEMKIND_DAX_KMEM** use case.
+
+**MEMKIND_DAX_KMEM_ALL**:
+This is the second variant where the allocation comes from the closest persistent memory NUMA node
```
Comment (pbalcer) [inline _posts/2019-12-31-memkind-dax-kmem.md]: where the memory

File: _posts/2019-12-31-memkind-dax-kmem.md [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+available at the time allocating.
+
+![memkind_dax_kmem_all](/assets/memkind_dax_kmem_all.gif)
+
+A similar situation to scenario presented in **MEMKIND_DAX_KMEM** except that when there is not enough free
+memory in Node 2 to satisfy the allocation request, an allocation pattern switch to Node 3. When available free space is
+exhausted in Node 3 - swap space is used.
+
+**MEMKIND_DAX_KMEM_PREFERRED**:
+This is the third variant where the allocation comes from the closest persistent memory NUMA node at the time
```
Comment (pbalcer) [inline _posts/2019-12-31-memkind-dax-kmem.md]: where the memory

File: _posts/2019-12-31-memkind-dax-kmem.md [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+of allocating. If there is not enough memory to satisfy the request the allocation will fall back on other memory
+NUMA nodes.
+
+![memkind_dax_kmem_preferred](/assets/memkind_dax_kmem_preferred.gif)
+
+Again the allocation starts from Node 2. When there is not enough free memory in Node 2, the allocation switches to other
+nodes in order of increasing distance from the preferred node based on information provided by the platform firmware:
```
Comment (pbalcer) [inline _posts/2019-12-31-memkind-dax-kmem.md]: in order to increase

File: _posts/2019-12-31-memkind-dax-kmem.md [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+Node 0, Node 3 and ending in Node 1. When available free space is
+exhausted in Node 1 - swap space is used.
+
+
+**Caveats**
+
+For **MEMKIND_DAX_KMEM_PREFERRED**, the allocation will not succeed if two or more
+persistent memory NUMA nodes are in the same shortest distance to the same CPU on which the process is eligible to run.
```
Comment (pbalcer) [inline _posts/2019-12-31-memkind-dax-kmem.md]: Eligible is an odd word to use here. "Permitted", "Allowed" ?

**Comment by michalbiesek** [comment]
> Since you have pictures here, it would be useful to have a rendered page. I checked, and michalbiesek.github.io is not configured.

Done

File: _posts/2019-12-31-memkind-dax-kmem.md (hunk 14 above)
Comment (michalbiesek) [reply _posts/2019-12-31-memkind-dax-kmem.md]: Done - just FYI i take this eligible from sched_setaffinity(2) manual page

File: _posts/2019-12-31-memkind-dax-kmem.md (hunk 13 above)
Comment (michalbiesek) [reply _posts/2019-12-31-memkind-dax-kmem.md]: Done

File: _posts/2019-12-31-memkind-dax-kmem.md (hunk 12 above)
Comment (michalbiesek) [reply _posts/2019-12-31-memkind-dax-kmem.md]: Done

File: _posts/2019-12-31-memkind-dax-kmem.md (hunk 11 above)
Comment (michalbiesek) [reply _posts/2019-12-31-memkind-dax-kmem.md]: Done

File: _posts/2019-12-31-memkind-dax-kmem.md (hunk 10 above)
Comment (michalbiesek) [reply _posts/2019-12-31-memkind-dax-kmem.md]: Done

File: _posts/2019-12-31-memkind-dax-kmem.md (hunk 9 above)
Comment (michalbiesek) [reply _posts/2019-12-31-memkind-dax-kmem.md]: Done

File: _posts/2019-12-31-memkind-dax-kmem.md (hunk 8 above)
Comment (michalbiesek) [reply _posts/2019-12-31-memkind-dax-kmem.md]: Done

File: _posts/2019-12-31-memkind-dax-kmem.md (hunk 7 above)
Comment (michalbiesek) [reply _posts/2019-12-31-memkind-dax-kmem.md]: Done

File: _posts/2019-12-31-memkind-dax-kmem.md (hunk 6 above)
Comment (michalbiesek) [reply _posts/2019-12-31-memkind-dax-kmem.md]: Done

File: _posts/2019-12-31-memkind-dax-kmem.md (hunk 5 above)
Comment (michalbiesek) [reply _posts/2019-12-31-memkind-dax-kmem.md]: Done

File: _posts/2019-12-31-memkind-dax-kmem.md (hunk 4 above)
Comment (michalbiesek) [reply _posts/2019-12-31-memkind-dax-kmem.md]: Done

File: _posts/2019-12-31-memkind-dax-kmem.md (hunk 3 above)
Comment (michalbiesek) [reply _posts/2019-12-31-memkind-dax-kmem.md]: Done

File: _posts/2019-12-31-memkind-dax-kmem.md (hunk 2 above)
Comment (michalbiesek) [reply _posts/2019-12-31-memkind-dax-kmem.md]: Done

File: _posts/2019-12-31-memkind-dax-kmem.md (hunk 1 above)
Comment (michalbiesek) [reply _posts/2019-12-31-memkind-dax-kmem.md]: Done

File: _posts/2019-12-31-memkind-dax-kmem.md (hunk 14 above)
Comment (pbalcer) [reply _posts/2019-12-31-memkind-dax-kmem.md]: If it's from the man page then leave it as eligible. It still sounds odd to me :P

**Comment by George330** [comment]
___
*[_posts/2019-12-31-memkind-dax-kmem.md, line 114 at r4](https://reviewable.io/reviews/pmem/pmem.github.io/120#-Lyn_p7YCePfEyhiOgeR:-Lyn_p7YCePfEyhiOgeS:b46g5xk) ([raw file](https://github.com/pmem/pmem.github.io/blob/18a3807f9d310438ac15be7841a8fc7c0aa9e6ab/_posts/2019-12-31-memkind-dax-kmem.md#L114)):*
> ```
>   see animations below.
> ```

File: _posts/2020-01-20-memkind-dax-kmem.md [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+as a hotplugged memory region (**KMEM DAX**).
+From the perspective of OS, persistent memory will be seen as a separate NUMA node(s).
+[libmemkind][memkind-release] API was extended to include new kinds that allow for automatic detection and allocation
+from these new persistent memory NUMA nodes.
+
+### Requirements
+
+<span>1.</span> Kernel 5.1 with **KMEM DAX** driver enabled:
+
+To configure proper driver installation run `nconfig` and enable driver.
```
Comment (marcinslusarz) [inline _posts/2020-01-20-memkind-dax-kmem.md]: You don't need to compile your own kernel (which this sentence only implies) if support for KMEM DAX is already compiled in, so this sentence should reflect that.

File: _posts/2020-01-20-memkind-dax-kmem.md [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+]
+{% endhighlight %}
+
+To [reconfigure][daxctl-reconfigure-device] DAX device from `devdax` mode to a `system-ram` mode:
+
+{% highlight console %}
+$ daxctl reconfigure-device dax1.0 --mode=system-ram
+{% endhighlight %}
+
+After this operation, persistent memory will be configured as a separate NUMA node and can be used as a volatile memory.
```
Comment (marcinslusarz) [inline _posts/2020-01-20-memkind-dax-kmem.md]: will be -> is?

File: _posts/2020-01-20-memkind-dax-kmem.md [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+For the example configuration below, persistent memory NUMA node is Node 3 (NUMA node to which does not belong any CPU):
```
Comment (marcinslusarz) [inline _posts/2020-01-20-memkind-dax-kmem.md]: NUMA node without any assigned CPU?

File: _posts/2020-01-20-memkind-dax-kmem.md [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+**Caveats**
+
+For **MEMKIND_DAX_KMEM_PREFERRED**, the allocation will not succeed if two or more
+persistent memory NUMA nodes are in the same shortest distance to the same CPU on which the process is eligible to run.
+A check on that eligibility is performed upon starting the application.
+
+### Environment variable
+
+**MEMKIND_DAX_KMEM_NODES** is an environment variable - a comma-separated list of NUMA nodes that
+are treated as persistent memory. This mechanism is dedicated to overwriting the automatic
```
Comment (marcinslusarz) [inline _posts/2020-01-20-memkind-dax-kmem.md]: overwriting -> override?

**Pull request merged**